### ISSUE TITLE
Objective class is confusing

### ISSUE THREAD

**Issue by ebenolson** [Lasagne/Lasagne#173]
In particular the argument `input_layer`, which I think would make more sense if renamed to `output_layer`

Here is the current docstring:

>- input_layer : a `Layer` whose output is the networks prediction
        given its input

While this technically does indicate that `input_layer` should be the output layer of the network, I think something like this would be better:

- output_layer: The `Layer` whose output will be compared to the target.

Also, is there a reason not to specify `target` in the constructor?

**Comment by benanne** [comment]
This is not quite a duplicate of #15, but there is quite some overlap (see my latest comment there).

The reason the argument is `input_layer` is because `Objective` used to be called `OutputLayer` in my galaxy challenge code, which Lasagne is largerly based on. Which made even less sense ;)

Regarding specifying `target` in the constructor, as I commented in the other issue we may just replace `Objective` classes with functions, so then this doesn't matter. But if we keep them as classes, I guess it would be useful to be able to set it in the constructor, for symmetry with `InputLayer`.

**Comment by benanne** [comment]
Thinking about this again, I'm not sure if we should add the abilty to specify the `target` in the constructor. The use case for `Objective` is to avoid having to manually declare Theano variables. So adding the ability to specify a custom target variable kind of defeats the point.

If you wish to use a custom target variable, you can just write something like `loss = T.mean(T.nnet.binary_crossentropy(x, t))` directly. No need to use an `Objective` for that.

Are there any other uses cases for `Objective` instances?

(Renaming `input_layer` to `output_layer` is a good idea regardless which I will implement when I address the additional remarks from #15.)

**Comment by ebenolson** [comment]
Sorry, I haven't tried to use `Objective` yet in my code so I'm a bit hazy on it. I think I suggested that change because looking at the current examples, it seemed like duplication could be reduced by putting those arguments in the constructor:

    loss_train = objective.get_loss(X_batch, target=y_batch)
    loss_eval = objective.get_loss(X_batch, target=y_batch,
                                   deterministic=True)

But it seems like the intended use is more like this?

    loss_train = objective.get_loss()
    loss_eval = objective.get_loss(deterministic=True)
    ...
    ...
    iter_train = theano.function(
        [batch_index], loss_train,
        updates=updates,
        givens={
            input_layer.input_var: dataset['X_train'][batch_slice],
            objective.target_var: dataset['y_train'][batch_slice],
        },
    )

**Comment by f0k** [comment]
> The use case for `Objective` is to avoid having to manually declare Theano variables.

That's only *one* use case. In my framework, the Theano variables are already there, so it should be possible to override `target_var`, as in the `InputLayer`.

> Are there any other uses cases for `Objective` instances?

Using `MaskedObjective` would be cleaner than writing `T.sum(T.nnet.binary_crossentropy(x, t) * weights) / weights.sum()` (although I'm not actually using that variant anyway).

**Comment by benanne** [comment]
> That's only *one* use case. In my framework, the Theano variables are already there, so it should be possible to override `target_var`, as in the `InputLayer`.

Well that's what I'm wondering about, what are the others? And is it really necessary to use `Objective` in any of them? To me it seems much cleaner to just specify a Theano expression for the loss in that case (or write a function for it, or whatever tickles your fancy).

> Using `MaskedObjective` would be cleaner than writing `T.sum(T.nnet.binary_crossentropy(x, t) * weights) / weights.sum()` (although I'm not actually using that variant anyway).

That's why we discussed refactoring the masking stuff though? Although to be honest I'm not sure how to do that in a clean way.

**Comment by f0k** [comment]
> That's why we discussed refactoring the masking stuff though?

Yes, true. In any case, I think that:
```python
def __init__(self, foo, bar, target_var=None):
    ...
    if target_var is None:
        target_var = something something
    self.target_var = target_var
```
is not much more complex than:
```python
def __init__(self, foo, bar):
    ...
    self.target_var = something something
```
Plus, it has the added benefit of documenting the mere existence of such a variable in a prominent place. So I'd vote for keeping it because it doesn't actually hurt. I'd still remove it from `get_loss()` to avoid the duplication. But just like @ebenolson, I haven't used the `Objective` class at all yet, so take this with a grain of salt.

**Comment by benanne** [comment]
The thing is, if we accept a replacement `input` variable in `get_loss()`, we should probably accept `target` there too.

When you create an `InputLayer`, you can also specify an input variable on construction and pass it to `get_output()`, so if we're going for symmetry we should support both of those here as well.

But my argument is that you only use `Objective` when you're using the internal input variable from `InputLayer` anyway (for symmetry reasons), so `Objective` does not need to support these 'overrides' (because you're not using them to begin with when you're using `Objective`).

To be honest I think we'd be better off if we rethought the entire API for this part of the library. It's just... not good :p

**Comment by f0k** [comment]
> But my argument is that you only use `Objective` when you're using the internal input variable from `InputLayer` anyway (for symmetry reasons)

Ah, okay, got you.

> To be honest I think we'd be better off if we rethought the entire API for this part of the library. It's just... not good :p

Yes! +1! We need to specify what this is actually supposed to solve, and it would be instructive to think about what we want the MNIST example to look like, because people will probably try to adopt whatever we do there.
Maybe the cleanest solution would be ditching the `Objective` class and just assembling the cost by hand, as that's the most transparent and flexible when it comes to combining different costs. Optionally, we could provide a helper function for doing a weighted/masked mean of some item-wise (per datapoint) or element-wise (per datapoint and feature) loss, but I think a major reason it was added was that `Objective` didn't support it. If we hadn't had `Objective` at that time, @Britefury would have just written down his masked loss directly and not have felt obliged to extend `Objective`.

**Comment by benanne** [comment]
Indeed, maybe we should just get rid of it for now and replace it with a bunch of utility functions. It would save us the headache of cleaning it up and committing to an API for it on short notice.

I think this is probably the best course of action, but let's wait a few days to hear what other people have to say.

**Comment by benanne** [comment]
What do you guys think about something like this? These functions are supposed to replace the existing `Objective` classes:

```python
def objective(input_layer, loss_function=mse, aggregation='mean',
              input=None, target=None, **kwargs):
    y = get_output(input_layer, input, **kwargs)
    # create a target variable the same shape as the output if none is
    # specified.
    t = y.type('target') if target is None else target

    losses = loss_function(y, t)

    if aggregation == 'mean':
        loss = losses.mean()
    elif aggregation == 'sum':
        loss = losses.sum()
    else:
        raise ValueError('aggregation must be \'mean\' or \'sum\', '
                         'got {0}'.format(aggregation))

    if target is None:
        return loss, t
    else:
        return loss


def masked_objective(input_layer, loss_function=mse, aggregation='mean',
                     input=None, target=None, mask=None, **kwargs):
    y = get_output(input_layer, input, **kwargs)
    # create a target variable the same shape as the output if none is
    # specified.
    t = y.type('target') if target is None else target
    losses = loss_function(y, t)
    m = losses.type('mask') if mask is None else mask
    masked_losses = losses * m

    if aggregation == 'mean':
        loss = masked_losses.mean()
    elif aggregation == 'sum':
        loss = masked_losses.sum()
    elif aggregation == 'normalized_sum':
        loss = masked_losses.sum() / mask.sum()
    else:
        raise ValueError('aggregation must be \'mean\', \'sum\' or '
                         '\'normalized_sum\','
                         ' got {0}'.format(aggregation)) 

    if target is None and mask is None:
        return loss
    else:
        ret = [loss]
        if target is not None:
            ret.append(t)
        if mask is not None:
            ret.append(m)
        return ret
```

As I've mentioned before, the main advantage to the `Objective` class for me is symmetry: I don't have to worry about the input variable because the `InputLayer` creates it for me, and I don't have to worry about the target variable because the `Objective` creates it for me.

So this alternative `objective` utility function is also able to create a target variable if you don't specify one, and will return it along with the loss expression if this is the case. Otherwise it will just return the loss expression (since you don't need an extra reference to the target variable you passed).

The classes don't have any real state, so I think these functions can do everything the classes can do.

Feedback is welcome. Once we settle in an interface I'll make a PR.

**Comment by f0k** [comment]
> renamed `input_layer` argument to `output_layer`

:+1: :+1: :+1:

> I don't have to worry about the input variable because the `InputLayer` creates it for me, and I don't have to worry about the target variable because the `Objective` creates it for me.

That's a valid reason to make `objective` a class, so you have the symmetry between `InputLayer.input_var` and `Objective.target_var` (plus `Objective.mask_var`).

> The classes don't have any real state, so I think these functions can do everything the classes can do.

The single advantage of the `Objective` class was that you could create an objective instance once, and then use it to get the loss both for training (with `deterministic=False`) and validation (with `deterministic=True`). With this proposal, it would need two calls of the `objective` function, and if you don't create the variables yourself, you would even end up with two sets of target + mask variables.

The disadvantage of the `Objective` class (for my use case) is that it entangles the `get_output` call, the `loss_function` call and the aggregation. This is still entangled now.

What about:
```python
def aggregate(loss, mode='mean'):
    masked_aggregate(loss, mask=False, mode=mode)
def masked_aggregate(loss, mask=None, mode='mean'):
    mask = ... if mask is None else mask
    if mask:
        loss *= mask
    if mode == 'mean':
        return loss.mean()
    ...

class Objective(object):
    def __init__(self, output_layer, loss_function, target_var=None, mode='mean'):
        ...
    def __call__(self, *args, **kwargs):
        return self.get_loss(*args, **kwargs)
    def get_loss(self, input=None, **kwargs):
        output = get_output(self.output_layer, input, **kwargs)
        loss = self.loss_function(output, self.target_var)
        return aggregate(loss, self.mode)
```
Not sure if it would be useful to pull out the *call `get_output` and `loss_function`* part as well. How would one go about auxiliary output layers as in GoogLeNet? What can we provide to make that easier?

**Comment by benanne** [comment]
I guess that is better in just about every way :) So I guess we're keeping the class for the reason you mentioned (hadn't really occurred to me).

Regarding multiple output layers, I don't think we necessarily need to support that in the `Objective` class. People can just use Theano code if they want to do fancy things. We shouldn't try to support every possible use case with one class / function / object / ..., it will just make the code very complicated to cater to 98% of use cases instead of 95%, and that extra 3% just isn't worth it imo.

**Comment by f0k** [comment]
> Regarding multiple output layers, I don't think we necessarily need to support that in the `Objective` class.

No, but maybe we can still provide some help. I don't think it's soo uncommon of a use case, and we could learn something about how to structure the `objective` module by thinking it through.

Without the module, and if the auxiliary losses use the same targets, it could be something like this:
```python
target_var = T.matrix()
main_loss = categorical_crossentropy(get_output(main_out), target_var)
aux_loss1 = categorical_crossentropy(get_output(aux_out1), target_var)
aux_loss2 = categorical_crossentropy(get_output(aux_out2), target_var)
total_loss = T.mean(mask_var * (main_loss + .5 * aux_loss1 + .5 * aux_loss2))
```
Ah wait, it must be:
```python
target_var = T.matrix()
expr_main_out, expr_aux_out1, expr_aux_out2 = get_output([main_out, aux_out1, aux_out2])
main_loss = categorical_crossentropy(expr_main_out, target_var)
aux_loss1 = categorical_crossentropy(expr_aux_out1, target_var)
aux_loss2 = categorical_crossentropy(expr_aux_out2, target_var)
total_loss = T.mean(mask_var * (main_loss + .5 * aux_loss1 + .5 * aux_loss2))
```
The last step would be slightly more readable/flexible with the `masked_aggregate` function. What we could still help with is (a) creating the target variable or variables, and (b) providing a way to create the same expression again with `deterministic=True`. Thinking about it, (b) is easily solved by the user by wrapping the statements in a function that takes `**kwargs` and passes them on to `get_output`. Is there any way we could help with (a)? Is there any reason we should?

> People can just use Theano code if they want to do fancy things.

I'm beginning to think that we should just ditch the `Objective` class because it makes the examples harder to follow, and harder to modify... Of course people can use Theano code, but the `Objective` class hides the three lines of Theano code they could use instead of the class, and that would be straightforward to extend to auxiliary losses. Hmm.

**Comment by benanne** [comment]
As I said, the reason I'm advocating for (a) is because of symmetry. I like not having to worry about creating appropriate variables. `InputLayer` creates the input variable for you, `Objective` takes care of the target variable. In many cases you can avoid using/importing `theano.tensor` completely this way.

But I agree that `Objective` hides some stuff that probably should not be hidden, and it mainly makes the code more opaque. So if we can't find a way to automate the variable creation in a clean way without the `Objective` class, I would still be okay with getting rid of it. Maybe we should just give up on this feature for now.

**Comment by f0k** [comment]
Current way:
```python
objective = Objective(layer, categorical_crossentropy)
target_var = objective.target_var
train_loss = objective.get_loss()
val_loss = objective.get_loss(deterministic=True)
```
Vanilla way:
```python
target = T.matrix()
train_loss = categorical_crossentropy(get_output(layer), target).mean()
val_loss = categorical_crossentropy(get_output(layer, deterministic=True), target).mean()
```
Another possibility (1):
```python
def make_target_var(output_layer, name='target', dtype=theano.config.floatX):
    ndim = len(output_layer.output_shape)
    return return T.TensorType(dtype, [False] * ndim)(name)
target = make_target_var(layer)
train_loss = categorical_crossentropy(get_output(layer), target).mean()
val_loss = categorical_crossentropy(get_output(layer, deterministic=True), target).mean()
```
Another possibility (2):
```python
def make_var_like(var, name=None):
    return var.type(name)
output = get_output(layer)
target = make_var_like(output, name='target')
train_loss = categorical_crossentropy(output, target).mean()
val_loss = categorical_crossentropy(get_output(layer, deterministic=True), target).mean()
```
Another possibility (3):
```python
def construct_loss(loss_function, output_layer, target_var=None, **kwargs):
    output = get_output(output_layer, **kwargs)
    if target_var is None:
        target_var = output.type(name='target')
        return loss_function(output, target_var), target_var
    else:
        return loss_function(output, target_var)
train_loss, target = construct_loss(categorical_crossentropy, layer).mean()
val_loss = construct_loss(categorical_crossentropy, layer, target, deterministic=True).mean()
```

All of the `.mean()` could be replaced by the `aggregate` function drafted above, but that's mostly useful if you do more complex things than taking the mean.

Looking at the different variants, I'm now pretty convinced that the one thing we should not hide is the `get_output` call. Doing so will trick people into introducing multiple `get_output` calls into their code without realizing, which will potentially run into #104.
And looking at the vanilla variant, I'm pretty convinced the `Objective` class does not only make it more opaque, but also makes it look like creating a loss expression was a complex task.

Would you be fine with just having a helper function that avoids the `T.matrix()` (or `T.tensor3()`, or `T.vector()`) call, as in variant (1) or (2)? For more symmetry to the `InputLayer`, we could also have a `TargetLayer` that you can put on top of every output layer of your network to create the variables for you, but nothing more. That wouldn't be a lot different to variant (1), though.

**Comment by benanne** [comment]
Yep, that sounds great! I think (2) is probably the cleanest option. Although it makes you wonder if we need a one-line helper function for this at all :)

By the way, I only discovered the `var.type(name)` thing when I was trying to come up with the 
`objective` function. The current `Objective` class doesn't even do this and just always creates a `T.matrix('target')`. Indeed, it couldn't, because it has to create the variable before `get_loss()` is called, so before it has any info about what the dimensionality should actually be.

So in summary let's kill `Objective`, the sooner the better :p

**Comment by f0k** [comment]
> Although it makes you wonder if we need a one-line helper function for this at all :)

That's right. The helper function reads better, though.

> I think (2) is probably the cleanest option.

I slightly tend towards (1), because it can create the variable independently of the `get_output` call. You only get into trouble if you need the same dtype and broadcasting pattern. That's an advantage of (2).

> The current `Objective` class doesn't even do this and just always creates a `T.matrix('target')`.

Uh. (But of course that could be fixed.)

Note that sometimes you don't event want the target variable to be of the same shape as the output. For example, for `categorical_crossentropy`, you might have the targets as a minibatch of integers instead of 1-hot vectors. Compared to `make_target_var`, the `make_var_like` variant communicates much clearer what it does, and if it does the right thing for your specific purpose. So I'm leaning towards (2) now.

> So in summary let's kill `Objective`, the sooner the better :p

Hooray! `Objective` is dead, long live `make_var_like` and `aggregate_loss`!

**Comment by benanne** [comment]
Yeah, I thought about the `categorical_crossentropy` case as well, but I didn't give much weight to it since you can also pass one-hot vectors (and we already have `lasagne.utils.one_hot`). But yeah, (2) wins by a narrow margin imo. Are you doing a PR for this or should I do it?

**Comment by f0k** [comment]
> But yeah, (2) wins by a narrow margin imo.

But I also kind of like the `TargetLayer` idea. It would allow you to clearly mark which layers are going to get some information from a teacher, and it would link the target variables directly to the output layers (or hidden layers) they belong to:
```python
class TargetLayer(Layer):
    def __init__(self, incoming, target_var=None, **kwargs):
        super(TargetLayer, self).__init__(incoming, **kwargs)
        if target_var is None:
            ndim = len(self.output_shape)
            name = ... # however we do it for the InputLayer
            target_var = T.TensorType(dtype, [False] * ndim)(name)
        self.target_var = target_var
    def get_output_for(self, input, **kwargs):
        return input
```
It would also allow some generic helper function that returns the output expressions for all `TargetLayer`s.
The only conceptual problem is that this is only useful for training, not for predicting, so it feels wrong to make it part of the architecture. We should probably keep it at variant (2) then.

> Are you doing a PR for this or should I do it?

I may be able to do it tomorrow. No more killing today.

**Comment by benanne** [comment]
Alright, no rush I guess! I'll focus on finishing the docs todo list and test coverage then.

**Comment by f0k** [comment]
Started working on it.
